### PR TITLE
[core] Insert guards to keep stack from growing without bound.

### PR DESCRIPTION
**Pull request by schweitzpgi**
This fixes a bug where classical variables may be allocated on the stack without bound. Instead, use the scope information to bound the lifetimes of classical variables and prevent a standard loop from accidentally exhausting the stack.

We haven't encountered this issue because we have focused on SCPs and totally unrolling all the loops. This patch is preventative maintenance.

Add test.

### CODE REVIEW

**Comment by schweitzpgi** [comment]
ping. needs a review.

File: lib/Optimizer/Dialect/CC/CCOps.cpp [hunk 1]
```diff
@@ -1428,6 +1428,28 @@ void cudaq::cc::ScopeOp::getSuccessorRegions(
...
 
+// If quantumAllocs, then just look for any allocate memory effect. Otherwise,
+// look for any allocate memory other than from the quake dialect.
+template <bool quantumAllocs>
+bool hasAllocation(Region &region) {
+  for (auto &block : region)
+    for (auto &op : block) {
+      if (auto mem = dyn_cast<MemoryEffectOpInterface>(op))
+        if (mem.hasEffect<MemoryEffects::Allocate>())
+          return quantumAllocs || !isa<quake::AllocaOp>(op);
```
[the remark this answers is not in the document]
Comment (bmhowe23) [reply lib/Optimizer/Dialect/CC/CCOps.cpp:1439]: Or more specifically, here's what I suggesting:
```suggestion
          if (quantumAllocs || !isa<quake::AllocaOp>(op))
            return true;
```

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply lib/Optimizer/Dialect/CC/CCOps.cpp:1439]: You're right. Thanks, nicely spotted.

**Review by bmhowe23 — APPROVED** [review]
👍

**Pull request merged**